regex2/challenges-11.test.js

```diff
@@ -0,0 +1,165 @@
+'use strict';
+
+// ------------------------------------------------------------------------------------------------
+// CHALLENGE 1
+//
+// Write a function named validatePin that uses a regular expression pattern to validate a PIN.
+//
+// If the PIN is four numerical digits long, return true. Otherwise, return false.
+// ------------------------------------------------------------------------------------------------
+
+const validatePin = (pin) => {
+  return /^\d\d\d\d$/.test(pin);
+};
+
+// ------------------------------------------------------------------------------------------------
+// CHALLENGE 2
+//
+// Write a function named findTagNames that iterates over an array of HTML strings
+// and uses a regular expression pattern to return the closing tags.
+//
+// For example, findTagNames(['<h1>Hello, world!</h1>', '<p>Welcome to my site</p>'])
+// returns ['/h1', '/p'], and findTagNames(['<div><h1>Hello, world!</h1></div>', '<p>Welcome to my site</p>'])
+// returns ['/h1', '/div', '/p'].
+// ------------------------------------------------------------------------------------------------
+
+const findTagNames = elements => {
+  let regex = /\/[a-z][a-z0-9]?[a-z]?/g;
+  return elements.reduce((acc, cur) => acc.concat(cur.match(regex)),[]);
+}
+
+// ------------------------------------------------------------------------------------------------
+// CHALLENGE 3
+//
+// Write a function named validateEmail that takes in an email address and validates it based
+// on several rules:
+//   - one word, or two words separated by a period, before the @ symbol
+//   - can contain numbers
+//   - can have any of the following top-level domains: .net, .com, or .org
+//   - no other special characters
+//   - no subdomains, ports, etc: must be of the form name@place.com, not name@sub.place.com:3000
+//
+// Return either true or false.
+//
+// Note: if you ever need to validate an email using a regex in practice, the Internet has the actual
+// regex you should use. It's many many lines long.
+// ------------------------------------------------------------------------------------------------
+
+const validateEmail = (email) => {
+  let regex = /^[a-zA-Z0-9]+.?[a-zA-Z0-9]+?@[a-zA-Z0-9]+\.(com)?(net)?(org)?$/;
```

Comment:
Same issue as we saw in class re: not working for a single character email. This will also (falsely) return true for `email@place.comnetorg`.